### PR TITLE
Give defaultLibrary semantic token modifier to items from standard library

### LINKED ISSUE #2155: Create scopes for std types.
It would be nice to have a separate scope for std types for more than just builtin primitive ones. 
Things like `Rc<T>`, `RefCell<T>` and traits like `Borrow<T>`. 
Xcode does this as it indicates standard library and frameworks with a different color than "project types".

### LINKED ISSUE #8999: Give `defaultLibrary` semantic token modifier to items from standard library
This should apply to all items from the `std`, `core`, `alloc`, `test` and `proc_macro` crates. Should these items also be given the `library` modifier? I’d argue they should, since they still are from a crate other than the current one.

### PR DESCRIPTION
**Pull request by lhvy**
Fixes #8999, fixes #2155

`builtInCrates` could be an alternate name to `defaultLibrary`, which one is better?

> P.S. I'm participating in [Hacktoberfest 2021](https://hacktoberfest.digitalocean.com/). If this PR is up to standard and merged, I'd appreciate if the `hacktoberfest-accepted` label could be added. Thanks!

### CODE REVIEW

File: crates/ide_db/src/helpers/famous_defs.rs [hunk 1]
```diff
@@ -68,6 +68,30 @@ impl FamousDefs<'_, '_> {
...
+        IntoIterator::into_iter([
+            self.std(),
+            self.core(),
+            self.alloc(),
+            self.test(),
+            self.proc_macro(),
+        ])
+        .filter_map(|it| it)
+        .collect()
+    }
```
Comment (Veykril) [inline crates/ide_db/src/helpers/famous_defs.rs:93]:
```suggestion
    pub fn builtin_crates(&self) -> impl Iterator<Item = Crate> {
        IntoIterator::into_iter([
            self.std(),
            self.core(),
            self.alloc(),
            self.test(),
            self.proc_macro(),
        ])
        .filter_map(|it| it)
    }
```
No need to allocate here

File: crates/rust-analyzer/src/to_proto.rs [hunk 2]
```diff
@@ -529,23 +529,24 @@ fn semantic_token_type_and_modifiers(
 
     for modifier in highlight.mods.iter() {
         let modifier = match modifier {
+            HlMod::Associated => continue,
+            HlMod::Async => semantic_tokens::ASYNC,
             HlMod::Attribute => semantic_tokens::ATTRIBUTE_MODIFIER,
+            HlMod::Callable => semantic_tokens::CALLABLE,
+            HlMod::Consuming => semantic_tokens::CONSUMING,
+            HlMod::ControlFlow => semantic_tokens::CONTROL_FLOW,
+            HlMod::DefaultLibrary => semantic_tokens::DEFAULT_LIBRARY,
```
Comment (Veykril) [inline crates/rust-analyzer/src/to_proto.rs:538]:
```suggestion
            HlMod::DefaultLibrary => lsp_types::SemanticTokenModifier::DEFAULT_LIBRARY,
```
This is pre-defined by the protocol so no need to define our own for this one

**Comment by Veykril** [comment]
We want to stick to `defaultLibrary` as that is one predefined by the protocol.
Could you also add a line to the comment here https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ide/src/syntax_highlighting.rs#L129-L146 for `defaultLibrary`?
Otherwise lgtm!

**Comment by lhvy** [comment]
Ok, that should've covered the review comments and tests are passing locally now. Let me know if anything else needs changing.

**Comment by Veykril** [comment]
Looks good to me! Thanks!
bors r+

**Comment by lnicola** [comment]
changelog feature (first contribution) set `defaultLibrary` semantic token on standard library items.

**Pull request merged**